broker-daemon/utils/basic-auth.js

```diff
@@ -0,0 +1,44 @@
+const { PublicError } = require('grpc-methods')
+
+/**
+ * @constant
+ * @type {Boolean}
+ * @default
+ */
+const BASIC_AUTH_DELIMITER = ':'
+
+/**
+ * Verifies a username and password from metadata sent from grpc-methods
+ *
+ * @see {@link https://github.com/sparkswap/grpc-methods}
+ * @param {Object} params - param object from grpc-methods
+ * @param {Object} metadata
+ * @param {Object} logger
+ * @param {Boolean} this.disableAuth
+ * @param {String} this.rpcUser
+ * @param {String} this.rpcPass
+ * @return {True} true - if the user is authenticated
+ * @return {Error} if the user is not authenticated
+ */
+function verify ({ metadata, logger }) {
+  if (this.disableAuth === true) return
+
+  // Example basic auth token: 'Basic YWRtaW46cGFzc3dvcmQ='
+  const { authorization: authToken } = metadata
+  const [scheme, base64Token] = authToken.split(' ')
+  logger.debug('Received auth token', { scheme, base64Token })
+
+  const rawCredentials = Buffer.from(base64Token, 'base64').toString()
+  const [username, password] = rawCredentials.split(BASIC_AUTH_DELIMITER)
+
+  logger.debug('Checking if credentials are valid between cli and broker')
+
+  // rpcUser and rpcPass are bound to context where #verify is used due to the way
+  // grpc-method handles the authorization middleware
+  if (username !== this.rpcUser || password !== this.rpcPass) {
```

Comment:
see previous comment - if we can avoid using implicit context I think we should.

Instead, we could do something like: `createAuth(user, pass)` that returns a function.

Reply:
ill have to think about this. The issue is that verify is called from within grpc-methods which has no information/data for these parameters (which is why we use bind above)

Reply:
you can instantiate within the RPC Server constructor and still pass a dumb method to grpc-methods.

```
   const auth = createAuth(user, pass)
   this.adminService = new AdminService(..., auth)
```